### PR TITLE
[Misc][LoRA] Support Rank Stabilized LoRA (RSLoRA)

### LINKED ISSUE #10798: [Bug]: LoRa adapter responses not matching peft/transformers response
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-48-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A40
Nvidia driver version: 550.127.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        52 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               96
On-line CPU(s) list:                  0-95
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Gold 6342 CPU @ 2.80GHz
CPU family:                           6
Model:                                106
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            2
Stepping:                             6
CPU max MHz:                          3500.0000
CPU min MHz:                          800.0000
BogoMIPS:                             5600.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            2.3 MiB (48 instances)
L1i cache:                            1.5 MiB (48 instances)
L2 cache:                             60 MiB (48 instances)
L3 cache:                             72 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-23,48-71
NUMA node1 CPU(s):                    24-47,72-95
Vulnerability Gather data sampling:   Mitigation; Microcode
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[uv] numpy==1.26.2
[uv] nvidia-cublas-cu12==12.1.3.1
[uv] nvidia-cuda-cupti-cu12==12.1.105
[uv] nvidia-cuda-nvrtc-cu12==12.1.105
[uv] nvidia-cuda-runtime-cu12==12.1.105
[uv] nvidia-cudnn-cu12==8.9.2.26
[uv] nvidia-cufft-cu12==11.0.2.54
[uv] nvidia-curand-cu12==10.3.2.106
[uv] nvidia-cusolver-cu12==11.4.5.107
[uv] nvidia-cusparse-cu12==12.1.0.106
[uv] nvidia-nccl-cu12==2.18.1
[uv] nvidia-nvjitlink-cu12==12.3.101
[uv] nvidia-nvtx-cu12==12.1.105
[uv] pyzmq==24.0.1
[uv] torch==2.1.1
[uv] torchaudio==2.1.1
[uv] torchvision==0.16.1
[uv] transformers==4.46.3
[uv] triton==2.1.0
[conda] Could not collect
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.6.4.post1
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      0-23,48-71      0               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

LD_LIBRARY_PATH=/workspace/ADVANCED-inference/lora-server/.venv/lib/python3.10/site-packages/cv2/../../lib64:
CUDA_MODULE_LOADING=LAZY
```

</details>

### Model Input Dumps

_No response_

### 🐛 Describe the bug

**Issue: LoRa adapter responses with vLLM do not match peft/transformer responses.**

The reproduction involves running inference using a) vllm, compared to b) running inference with transformers and peft. I have run both on A40 machines on runpod.

# vLLM approach

## vLLM server startup with statically loaded LoRa
```
ADAPTER_REPO="Trelis/Qwen2.5-7B-Instruct-touch-rugby-1"
ADAPTER_PATH=$(python3 -c "from huggingface_hub import snapshot_download; print(snapshot_download('${ADAPTER_REPO}', ignore_patterns=['model-*', 'pytorch_model*', 'tf_model*', 'flax_model*']))")

# Start server with LoRA pre-loaded
vllm serve Qwen/Qwen2.5-7B-Instruct \
    --max-model-len 8192 \
    --enable-lora \
    --max-lora-rank 8 \
    --port 8000 \
    --lora-modules "{\"name\": \"touch-rugby-1\", \"path\": \"${ADAPTER_PATH}\", \"base_model_name\": \"Qwen/Qwen2.5-7B-Instruct\"}"
```

## vLLM Script to call the server endpoint, aka `vllm_replication.py`
```
import requests
import json
from typing import List, Dict, Any

def send_chat_completion(
    messages: List[Dict[str, str]],
    model: str,
    base_url: str = "http://localhost:8000"
) -> Dict[str, Any]:
    """Send a chat completion request to the vLLM server."""
    try:
        response = requests.post(
            f"{base_url}/v1/chat/completions",
            headers={"Content-Type": "application/json"},
            json={
                "model": model,
                "messages": messages,
                "max_tokens": 100,
                "temperature": 0.01,
            }
        )
        response.raise_for_status()  # Raise an error for bad status codes
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f"Error making request: {e}")
        print(f"Response content: {response.text if 'response' in locals() else 'No response'}")
        return None

def main():
    # Configuration
    base_model = "Qwen/Qwen2.5-7B-Instruct"
    adapter_name = "touch-rugby-1"  # This matches the name in your vLLM server command
    
    # Test prompt (same as in your transformers example)
    messages = [{
        "role": "user",
        "content": "How many players are on the field on each team at the start of a drop-off?"
    }]

    # Generate with base model
    print("\nBase Model Response:")
    base_response = send_chat_completion(messages, base_model)
    if base_response and 'choices' in base_response:
        print(f"Base Model Output: {base_response['choices'][0]['message']['content']}")
    else:
        print("Failed to get base model response")

    # Generate with adapter
    print("\nAdapter Model Response:")
    adapter_response = send_chat_completion(messages, adapter_name)
    if adapter_response and 'choices' in adapter_response:
        print(f"Adapter Output: {adapter_response['choices'][0]['message']['content']}")
    else:
        print("Failed to get adapter response")

if __name__ == "__main__":
    main()
```

# Transformers / PEFT script
```
# Install required packages if needed
!pip install transformers peft accelerate -qU

import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import PeftModel

# Load the base model and tokenizer
base_model_id = "Qwen/Qwen2.5-7B-Instruct"
adapter_id = "Trelis/Qwen2.5-7B-Instruct-touch-rugby-1"

tokenizer = AutoTokenizer.from_pretrained(base_model_id, trust_remote_code=True)
base_model = AutoModelForCausalLM.from_pretrained(
    base_model_id,
    torch_dtype=torch.bfloat16,
    device_map="auto",
    trust_remote_code=True
)

# Test prompt
prompt = "How many players are on the field on each team at the start of a drop-off?"

# Function to generate response
def generate_response(model, prompt):
    messages = [
        {"role": "user", "content": prompt}
    ]

    text = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
    
    inputs = tokenizer(text, return_tensors="pt").to(model.device)
    
    outputs = model.generate(
        **inputs,
        max_new_tokens=100,
        temperature=0.01,
        do_sample=True,
    )
    
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    return response

# Generate with base model
print("Base Model Response:")
base_response = generate_response(base_model, prompt)
print(base_response)

# Load adapter and generate
print("\nLoading adapter...")
adapter_model = PeftModel.from_pretrained(base_model, adapter_id)

print("\nAdapter Model Response:")
adapter_response = generate_response(adapter_model, prompt)
print(adapter_response)

# Free up memory
del base_model
del adapter_model
torch.cuda.empty_cache()
```

# Results / Output

```
uv run vllm-replication.py

Base Model Response:
Base Model Output: It seems there might be a bit of confusion in your question. The term "drop-off" is not commonly used in sports, especially not in team sports like football, basketball, or soccer. Could you please clarify which sport you are referring to? 

For example:
- In soccer (football), there are 11 players on the field for each team at the start of a match.
- In American football, there are 11 players on the field for each team at the start of

Adapter Model Response:
Adapter Output: In the context of Australian rules football, which is the sport referred to as "drop-off," each team starts with 18 players on the field at the beginning of the game. This number of players is standard for a match and includes both forwards, midfielders, and defenders.
```

```
Base Model Response:
system
You are Qwen, created by Alibaba Cloud. You are a helpful assistant.
user
How many players are on the field on each team at the start of a drop-off?
assistant
It seems there might be some confusion in your question. The term "drop-off" is not commonly used in sports to describe the number of players on a field or court. Could you please clarify which sport you are referring to? 

For example:
- In soccer (football), there are 11 players on the field for each team.
- In American football, there are typically 11 players on the field for each team at the start of a play.
- In basketball, there are

Loading adapter...

Adapter Model Response:
system
You are Qwen, created by Alibaba Cloud. You are a helpful assistant.
user
How many players are on the field on each team at the start of a drop-off?
assistant
To determine how many players are on the field at the start of a drop-off, we can follow these steps:

1. **Understand the Composition of a Team**: Each team consists of 14 players, including the Interchange.

2. **Interchange Rules**: The Interchange is allowed to enter and leave the field during normal play without a Change of Possession (COP).

3. **Drop-Off Procedure**: A Drop-Off occurs when one team has fewer than six (
```

# Additional Notes
- Base models are loaded in bfloat16 in both cases
- The very same adapters and base models are used from Huggingface
- The adapter is public and you can check it [here](https://huggingface.co/Trelis/Qwen2.5-7B-Instruct-touch-rugby-1)
- I also have tried with the same fine-tuning approach on Llama 3.1 8B (with ranks of 8 and also 32).
- I intentionally used a low rank of 8 here in case somehow that gave issues.

# Questions
- Are there any known issues with loading safetensors? I don't see anything in the issues?
- Are there limits on what modules can be trained? I only trained linear layers, which should be fine:
 ```
"target_modules": [
  | "gate_proj",
  | "k_proj",
  | "v_proj",
  | "down_proj",
  | "q_proj",
  | "o_proj",
  | "up_proj"
  | ],

```

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by JohnGiorgi**
vLLM will produce slightly incorrect outputs if a user provides a LoRA adapter trained with [Rank Stabilized LoRA](https://huggingface.co/papers/2312.03732). I _think_ the only change needed is to detect when a LoRA adapter has set `use_rslora` and then scale the `lora_alpha` accordingly, i.e.

`lora_alpha = config["lora_alpha"] * math.sqrt(rank) if config["use_rslora"] else config["lora_alpha"]`

This one-line PR does exactly that!

I confirmed for my own problem, where adapters are trained with RS-LoRA, this change leads to parity in model outputs between vanilla HuggingFace and vLLM, and without it, the outputs from vLLM are significantly different (and in my estimation, worse). Unfortunately I can't share this model but I could find or create a public example if need be. However, given that all RS LoRA does is scale `alpha` by `sqrt(r)` I think this change is de-risked.

I'd be happy to add a unit test with a little guidance!

FIX: https://github.com/vllm-project/vllm/issues/10798

### CODE REVIEW

**Comment by JohnGiorgi** [comment]
This fix works for my setup, but on second inspection it's a little awkward:

- It will only work when LoRA checkpoints are loaded with `LoRAModel.from_local_checkpoint` (not `LoRAModel.from_lora_tensors`)
- It is technically performing `(alpha * sqrt(r)) / r`, instead of simply `alpha / sqrt(r)`,  because it applies the `*` in `LoRAModel.from_local_checkpoint` here:

https://github.com/vllm-project/vllm/blob/c66c7f86aca956014d9ec6cc7a3e6001037e4655/vllm/lora/models.py#L371

and the `/` in `LoRALayerWeights` here:

https://github.com/vllm-project/vllm/blob/c66c7f86aca956014d9ec6cc7a3e6001037e4655/vllm/lora/lora.py#L30-L33

I think the solution to both problems is to apply the RS LoRA scaling once, in `LoRALayerWeights.__init__`. However this class currently has no way of knowing if `use_rslora=True`. It would be straightforward to provide that from `LoRAModel.from_local_checkpoint`, but I am not sure how to pass it from `LoRAModel.from_lora_tensors`

**Comment by JohnGiorgi** [comment]
Ping

**Comment by jeejeelee** [comment]
> Is there anyone who could take a look at this? AFAICT models fine-tuned with RSLoRA still aren't working correctly in vLLM due to this bug (cc @Yard1)

Thanks  for your contribution. 
Sorry for the delayed response. Would you still be interested in completing this feature? If so, please let me know.

**Comment by JohnGiorgi** [comment]
> > Is there anyone who could take a look at this? AFAICT models fine-tuned with RSLoRA still aren't working correctly in vLLM due to this bug (cc @Yard1)
> 
> Thanks for your contribution. Sorry for the delayed response. Would you still be interested in completing this feature? If so, please let me know.

Yes! Happy to help get this over the finish line, just had a couple of concerns I listed above, mainly:

- Solution will only work when LoRA checkpoints are loaded with `LoRAModel.from_local_checkpoint` (not `LoRAModel.from_lora_tensors`) --> Is this okay?
- It is technically performing `(alpha * sqrt(r)) / r`, instead of simply `alpha / sqrt(r)` (both give you the same answer)

**Comment by jeejeelee** [comment]
I really appreciate your response.  

For RSLoRA specifically, your concerns are all okay . However, considering our future plans to support DoRA(see: https://github.com/vllm-project/vllm/issues/10849), I think we need to give this more thought. I suggest we could implement a class that stores some lora information(e.g. rank, use_rslora,lora_alpha and so on) and pass it to `from_lora_tensors`. This might be a more suitable approach

**Comment by RonanKMcGovern** [comment]
Is it a valid fix to change the alpha value on the saved Lora to

alpha_new = alpha * sqrt(r)

Should be mathematically equivalent when the Lora is loaded then right?

**Comment by jeejeelee** [comment]
> Is it a valid fix to change the alpha value on the saved Lora to
> 
> alpha_new = alpha * sqrt(r)
> 
> Should be mathematically equivalent when the Lora is loaded then right?
> 
> [Perhaps I am wrong but it seems an odd convention for the alpha or r to be applied at inference time. They are really learning rate hyperparameters, so the convention of rescaling delta W instead of just rescaling the adapter learning rate seems cumbersome. Not that this library has any say on this.]

THB, I haven't looked into this  implementation yet. We don't have control over how the LoRA weights are saved. What we can do is to align with PEFT.

**Comment by RonanKMcGovern** [comment]
Confirming that a hacky fix is to update the alpha in the adapter config as I described above:
alpha_new = sqrt(rank) * alpha_old

This will result in the same generation as peft.

The better fix would indeed to read in the rank and do that scaling if use_rslora is true.

**Comment by jeejeelee** [comment]
@JohnGiorgi  FYI: https://github.com/vllm-project/vllm/pull/11003 has been merged. I think  we can modify some code based on this to complete this PR

**Comment by JohnGiorgi** [comment]
Thanks @jeejeelee, hoping to dig into this tomorrow!

**Comment by nielsrolf** [comment]
Would be really cool if this gets merged soon! :)

**Comment by JohnGiorgi** [comment]
@jeejeelee Whoops, sorry got pulled away on something else. Will do this today! Is the fix just modifying my one-line change to reference `peft_helper.use_rslora`?

File: vllm/lora/peft_helper.py [hunk 1]
```diff
@@ -38,6 +38,8 @@ def _validate_features(self):
 
     def __post_init__(self):
         self._validate_features()
+        if self.use_rslora:
+            self.vllm_scaling_factor = self.lora_alpha / math.sqrt(self.r)
         if self.context_length:
             if self.vllm_max_position_embeddings is None:
                 self.vllm_max_position_embeddings = self.context_length
```
Comment (JohnGiorgi) [inline vllm/lora/peft_helper.py:45]: @jeejeelee I think this is the cleanest way to support the RSLoRA scaling logic! My main outstanding worry is I am not sure how the scaling applied for RSLoRA and the scaling applied when `self.context_length` is provided interact.

Comment (jeejeelee) [reply vllm/lora/peft_helper.py:45]: > My main outstanding worry is I am not sure how the scaling applied for rsLoRA and the scaling applied when self.context_length is provided should interact.

IMHO, These two `scaling` are distinct and operate independently of each other.

We shouldn't use `vllm_scaling_factor `- instead, we can add a variable called `scaling`

Comment (JohnGiorgi) [reply vllm/lora/peft_helper.py:45]: Fine by me, but its gets a little confusing as `LoRAModel` already defines `scaling_factor` and uses it for long context support:

```python
class LoRAModel(AdapterModel):
    """A LoRA fine-tuned model."""

    def __init__(
        self,
        lora_model_id: int,
        rank: int,
        loras: Dict[str, LoRALayerWeights],
        scaling_factor: Optional[float] = None,
    ) -> None:
        """
        Args:
            lora_model_id: The integer id for the lora model.
            rank: lora rank.
            loras: module name -> weights for lora-replaced layers.
            scaling_factor: Scaling factor to support long context lora model.
                None if the lora is not tuned for long context support.
```

Is there an argument to sync `LoRAModel` and `PEFTHelper` so each has a `scaling_factor` and `vllm_scaling_factor` argument?

Comment (jeejeelee) [reply vllm/lora/peft_helper.py:45]: `long_context_scaling_factor ` looks reasonable. Let me verify this tomorrow and confirm

Comment (jeejeelee) [reply vllm/lora/peft_helper.py]: I have implemented the related logic, please check if it's reasonable

**Comment by AIlaowong** [comment]
But v.0.6.5 is not supported anymore
error_msg:
ValueError: vLLM does not yet support RSLoRA.
image

File: vllm/lora/peft_helper.py (hunk 1 above)
Comment (JohnGiorgi) [reply vllm/lora/peft_helper.py]: This looks good to me @jeejeelee! Thanks for adding tests

Comment (jeejeelee) [reply vllm/lora/peft_helper.py]: Could you please sync with the main branch?

Comment (JohnGiorgi) [reply vllm/lora/peft_helper.py]: Done. It is just the DCO check that is failing but I am unsure how to fix, it suggests a rebase but I don't want to do that as there's multiple commit authors on this branch

**Review by jeejeelee — APPROVED** [review]
@JohnGiorgi  Thank you for your contribution and patience.

File: vllm/lora/peft_helper.py (hunk 1 above)
Comment (jeejeelee) [reply vllm/lora/peft_helper.py]: There is a known issue in the current lora test. Let's check if other lora tests can pass first. If they do, we can consider force merging

**Comment by JohnGiorgi** [comment]
> @JohnGiorgi Thank you for your contribution and patience.

Woohoo! Thanks for all the guidance

**Pull request merged**